Feature: BB-124 replace zookeeper by mongo GET in notification extension

welansari opened the pull request:
Issue: [BB-124](https://scality.atlassian.net/browse/BB-124)

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/backbeat/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-124 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.3.16`


Please check the `Fix Version/s` of BB-124, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
feature: BB-124 Replace zookeeper in NotificationConfigManager with Mongo

Bucket notification configurations are not initialized at startup, the configurations
are cached once retreived, and then get invalidated using a MongoDB change stream.

The configurations no longer need to be updated by hand, as we get them directly from
the metastore collection in MongoDB.

Commit:
feature: BB-124 removed zookeeper from the notification extension

Bucket entries are now ignored as we no longer update the configuration manually

Functions that process the object/kafka entries become asynchronous as now we use
MongoDB calls

Commit:
feature: BB-124 adapting QueuePopulator & task runners to the new changes in the notification extension

Also updated notification config & constants

Commit:
feature: BB-124 Adapting the logReader to the async notification filter

The filter function of the notification extension now uses MongoDB
queries instead of zookeeper which makes it asynchronous. We have to
wait for all the entries to be processed by it before emptying the batch

Commit:
feature: BB-124 remove zookeeper from notification part of crr

Replacing zookeeper by MongoDB when initializing the NotificationConfigManager

Fixed small bug, where the versionId of the object is kept and displayed in the
object's key.

jbertran commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,436 +1,216 @@
+    /**
+     * Handler for the change stream "change" event.
+     * Invalidates cached bucket configs based on the change.
+     * @param {Object} change Change stream change object
```
We may as well leverage the partial typing provided by the mongodb driver - this object is a [ChangeStreamDocument](https://mongodb.github.io/node-mongodb-native/4.5/interfaces/ChangeStreamDocument.html).

jbertran commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,436 +1,216 @@
-            return this._callbackHandler(cb, err);
+                break;
+            default:
+                this._logger.debug('Unsupported change stream event', {
```
Nitpick: `Skipping unsupported ... ` so we understand that this is an expected non-error case.

jbertran commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,436 +1,216 @@
-                    this._configs.set(bucket, configObject);
-                }
-                return next();
+                this._setMetastoreChangeStream(() => {});
```
Shouldn't we just clear listeners on the changeStream, rather than change the callback?

welansari replied:
i agree that we should clear listeners, but i think we should also try restarting the stream if possible
An error at the change stream level will not cause the NotificationConfigManager to malfunction in the sense that the `getConfig()` function will still work. So if we just close the change stream when getting an error, the config in cache will never get invalidated, and the extension would still receive and process notification events

jbertran commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,436 +1,216 @@
-    getConfig(bucket) {
-        return this._configs.get(bucket);
+    setup(cb) {
+        async.series([
```
Does this need to be async? Could just be setup/setmetastore wrapped in a try/except

jbertran commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -36,89 +34,10 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
     _isBucketEntry(bucket, key) {
-        return ((bucket.toLowerCase() === notifConstants.bucketMetastore && key)
+        return ((bucket.toLowerCase() === notifConstants.bucketMetastore && !!key)
             || key === undefined);
```
Wish we had TS here - that's a good candidate for a literal type

jbertran commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -128,18 +47,18 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
             let eventType
                 = value[eventMessageProperty.eventType];
-            if (eventType === undefined && type === 'del') {
+            if (eventType === undefined && type === 'delete') {
```
Where is this change from? (hey, another TS literal type waiting to happen)

welansari replied:
I'm not actually sure on when or why it got changed, but you can see the part of code that sets the type of the entries  here 

https://github.com/scality/Arsenal/blob/019907e2ab841acc70a32331fec77d99efc4265c/lib/storage/metadata/mongoclient/ListRecordStream.js#L120-L126

jbertran commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -178,9 +97,9 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
      *
      * @param {Object} entry - log entry
-     * @return {undefined}
+     * @return {Promise|undefined} promise|undefiend
```
Typo

jbertran commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -189,15 +108,12 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
+        // ignore bucket operations, mpu's or if the entry has no bucket
         if (!bucket || bucket === usersBucket ||
-            (key && key.startsWith(mpuBucketPrefix))) {
+            (key && key.startsWith(mpuBucketPrefix)) ||
```
For legibility, this could become a `.some` call

jbertran reviewed (commented):
A few nitpicks. Let's sync tomorrow to set my brain back on track

jonathan-gramain commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,436 +1,216 @@
     logger: joi.object().required(),
 }).required();
 
+const MAX_CACHED_ENTRIES = 1000;
```
This should be configurable IMHO since it's a tunable, maybe via an environment variable

jonathan-gramain commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,436 +1,216 @@
- * @classdesc Manages bucket notification configurations in zookeeper, maintains
- * configurations across multiple managers.
+ * @classdesc Manages bucket notification configurations, the configurations
+ * are directly retreived from the metastore, and are locally cached. Cache
```
```suggestion
 * are directly retrieved from the metastore, and are locally cached. Cache
```

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-124 contains:


* `8.3.16`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.0`


Please check the `Fix Version/s` of BB-124, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jbertran reviewed (approved):
LGTM!

jonathan-gramain commented on lib/queuePopulator/LogReader.js:
```diff
@@ -407,25 +416,29 @@ class LogReader {
+             * waiting for all sync and async extentions
+             * to finish processing before emptying the batch
+             */
+            async.each(record.entries, async entry => {
```
It may be important to process the entries in series (as is, only call the next async `_processLogEntry` when the previous one completes), in which case using `async.eachSeries` is needed. Otherwise, it could still be nicer to use `async.eachLimit` to limit the parallelism.

Kerkesni commented:
/approve

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-124 contains:


* `8.4.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.4.1`


Please check the `Fix Version/s` of BB-124, or the target
branch of this pull request.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.2`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

welansari commented:
/approve

Kerkesni commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **approve**

bertranddemiddelaer commented:
/wait

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.4`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/8.2`
* `development/8.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.4`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/8.2`
* `development/8.3`



Please check the status of the associated issue BB-124.

Goodbye kerkesni.
